「kaminariを使ってページング処理を実装する」課題の提出

wata00913 opened the pull request:
確認をよろしくお願いします！🙇‍♂️

## 補足説明
- 歓迎要件「kaminariの"First"や"Last"のような表示を日本語化する」を対応しました。
- デフォルトで表示される以下の項目は削除しました。（Google検索結果で表示されるページングのレイアウトに近づけたかったためです。）
  - First
  - Last

wata00913 committed:
gemのkaminariを追加

wata00913 committed:
kaminariの設定ファイルを追加

wata00913 committed:
一覧画面のページング処理を追加

wata00913 committed:
ページングのテンプレートを追加

後ほどページングの見た目を変更するため

wata00913 committed:
ページング構成の変更とそれに伴う不要なファイルを削除

wata00913 committed:
ページングのロケール対応

wata00913 committed:
デフォルトの1ページ当たりの項目数を変更

maedana commented on config/initializers/kaminari_config.rb:
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+Kaminari.configure do |config|
+ config.default_per_page = 2
```
凄く細かいんですが、indentが1文字スペースになっちゃってますかね

wata00913 replied:
インデントの修正を行いました！

fjord-rubocopのパスは通ったので気づきませんでした。
こういったミスをしないように工夫しないと駄目ですね😅
インデントは、vimのLSPクライアントかvimの設定で保存時に自動フォーマットしようと思います💪

maedana commented on config/locales/translation_ja.yml:
```diff
@@ -9,3 +9,7 @@ ja:
+  views:
+    pagination:
+      next: 次へ
+      previous: 前へ
```
次へ、前へだけじゃちょっと足りないかなーと思います。
https://github.com/kaminari/kaminari#i18n-and-labels

wata00913 replied:
見返すと学習面、使い勝手共に物足りないと思いました。😅
フィヨルドブートキャンプアプリを見ると、きちんと「<<」、「>>」、「...」があって、「次へ」、「前へ」だけだと使い勝手も悪いですよね。

デフォルトで用意されてた「最初へ」、「最後へ」、「...」を再度追加しました！

maedana commented on config/locales/translation_en.yml:
```diff
@@ -9,3 +9,7 @@ en:
+  views:
+    pagination:
+      next: next
+      previous: prev
```
英語は変更する理由がなければデフォルトでいいので不要ですね

wata00913 replied:
英語を変更する理由が無かったのでデフォルトを使用しました！
また、[上のレビュー](https://github.com/wata00913/fjord-books_app/pull/2#discussion_r939896220)で追加した「最初へ」、「最後へ」、「...」も同様にデフォルトを使用しました。

maedana commented on app/views/kaminari/_next_page.html.erb:
```diff
@@ -0,0 +1,11 @@
+-%>
+<span class="next">
+  <%= link_to_unless current_page.last?, t('views.pagination.next').html_safe, url, rel: 'next', remote: remote %>
+</span>
```
このへんのkaminariのpartialは何故コミットしてあるんでしょう?

wata00913 replied:
YAGNI原則に違反してしまいました😓
「前へ」、「次へ」のテンプレートは後々変更するだろうと思って追加しました。
結局YAGNI原則通り、カスタマイズしなかったため、kaminariのテンプレート生成は不要でした。

なので、対応としては、アプリ内にあるkaminariテンプレートを全削除しました。

wata00913 committed:
インデントの修正

wata00913 committed:
ページングの英訳辞書を削除。ページングの日本語英訳をデフォルトの英語に合わせる

英語の翻訳はgem内のデフォルト翻訳と差分がないので不要。
また、英語翻訳では大なり、小なりの記号が含まれるので日本語もそれに合わせた。

wata00913 committed:
不要なファイルの削除

現状では生成したテンプレートファイルを編集してないので削除した。

wata00913 committed:
ページングの「最初へ」「最後へ」「三点リーダー(truncate)」を追加、kaminariで作成したテンプレートを削除。

ページングの「最初へ」「最後へ」を_paginator.erb.htmlに追加すると、gem内のテンプレートと差分がないため不要となった。

wata00913 committed:
表示ページの左右2ページ分のみ表示するように修正

ただし、表示対象の両端ページが以下の場合、例外的に最初または最終ページも表示される
- 表示ページの左端が２ページの場合、1ページも表示。
- 表示ページの右端が最終ページ直前の場合、最終ページも表示。

wata00913 commented:
ご指摘内容を修正しましたので、再確認をよろしくお願いします🙇‍♂️
すみません、色々書きたいことを書いたら長文になりました😅

## 今回の再提出で追加したこと
### 背景
レビューでの指摘箇所を修正をしましたが、デフォルトのものを使うだけなので物足りなさを感じました。
そこで、ページングの表示内容を少し変えてみました。

### 追加内容
ページングの表示内容の要件は以下になります。
※ 表示内容の具体例を挙げますが、現在のページ番号nを`[n]`と表現してます。

- ページの表示数は最大5件（但し、一部のページは例外）
- ページの続きがある場合は「...」を表示
- 現在のページを中心に左右2件ずつ表示
  - 左右いづれかの表示件数が2件に満たない場合は、反対側の件数を増やし最大5件表示を維持する
    - 例えば現在ページが2の場合、表示内容は`1 [2] 3 4 5 ...`

以下に現在のページ番号ごとに幾つか具体例を挙げます。
- `[1] 2 3 4 5 ... `
- `1 [2] 3 4 5 ... `
- ` ... 5 6 7 8 [9]`
- ` ... 5 6 7 [8] 9`
- ` ...3 4 [5] 6 7 ...`



### 例外について
現在ページの番号によっては、上記の要件を満たさない場合があります。
- 左右両端のページ番号に隣接するページ番号が先頭、または最終の場合
  - 先頭または最終ページ番号も表示される。

以下に例外のケースが発生する具体例を挙げます。
※今回のエビデンスでは、最終ページ番号を`9`とします。

- `1 2 3 [4] 5 6 ...`
- ` ...4 5 [6] 7 8 9`

### 例外の調査
例外に対処するために以下の方法を試みましたが、解消できてません。
- [kaminariのオプション](https://github.com/kaminari/kaminari#general-configuration-options)を参照し、解消できるオプションを探す

ソースコードを読んで原因調査しようとしましたが、時間がかかり過ぎて断念しました。

### 実装方法
`paginate`の引数オプション`window`を使いました。

maedana commented:
> 例外

これぐらいはいいかなーと思います。僕もちょっと解消方法がわかりません。仕様、といえば仕様だとも思います。
